### PR TITLE
Added a flatten module

### PR DESCRIPTION
**Pull request by Chillee**
https://github.com/pytorch/pytorch/issues/2118

I'm not sure I'm doing it correctly, so I'll add tests if we decide that it's roughly correct.

### CODE REVIEW

File: torch/nn/modules/flatten.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        - Output: :math:`(N, product of dims)`. Output is of the same shape as input
+
+    Examples::
+        >>> m = nn.Sequential(
+        >>>     nn.Conv2d(1, 32, 5, 1, 1),
+        >>>     nn.Flatten()
+        >>> )
+    """
+    @weak_script_method
+    def forward(self, input, start_dim=1, end_dim=-1):
```
Comment (SsnL) [inline torch/nn/modules/flatten.py]: the dim specification should be in the `__init__` ctor.

**Review by ssnl — CHANGES_REQUESTED** [review]
also need a test

**Review by fmassa — CHANGES_REQUESTED** [review]
This looks good, thanks!

Can you also add an entry in the documentation in https://github.com/pytorch/pytorch/blob/master/docs/source/nn.rst

File: torch/nn/modules/flatten.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+class Flatten(Module):
+    r"""
+    Flattens a contiguous range of dims into a tensor. For use with nn.Sequential.
+    Args:
+        start_dim: first dim to flatten
+        end_dim: last dim to flatten
+
+    Shape:
+        - Input: :math:`(N, *dims)
+        - Output: :math:`(N, product of dims)`. Output is of the same shape as input
```
Comment (SsnL) [inline torch/nn/modules/flatten.py]: This is inaccurate because the output shape depends on the arguments.

**Comment by Chillee** [comment]
Done. Hopefully I added it to the `rst` file properly...

File: torch/nn/modules/flatten.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+from .module import Module
+from ..._jit_internal import weak_module, weak_script_method
+
+
+@weak_module
+class Flatten(Module):
+    r"""
+    Flattens a contiguous range of dims into a tensor. For use with nn.Sequential.
```
Comment (SsnL) [inline torch/nn/modules/flatten.py]: `` :class:`~torch.nn.Sequential` ``

File: torch/nn/modules/flatten.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    Args:
+        start_dim: first dim to flatten
+        end_dim: last dim to flatten
+
+    Shape:
+        - Input: :math:`(N, *dims)
+        - Output: :math:`(N, product of dims)`
```
Comment (SsnL) [inline torch/nn/modules/flatten.py]: This is still inaccurate...

Comment (Chillee) [reply torch/nn/modules/flatten.py]: It's not entirely clear to me what's desired here. If we want to be completely accurate we could do `(N, \prod_{i=\text{start\_dim}}^{\text{end\_dim+1}} \text{dims[i]}, *dims[end_dim+1:])`, which renders like
image.

I'd also appreciate other suggestions on how to write this.

Comment (SsnL) [reply torch/nn/modules/flatten.py]: Even mentioning `N` is not quite right. The default arguments works for batched tensors, but it doesn't have to be batched. Since you already mentioned what the two dims are, instead of saying the general case, can you just say that explicitly, for the default arguments, what the input and output shapes are?

Comment (Chillee) [reply torch/nn/modules/flatten.py]: Sorry I'm not sure I understand your comment. If I were to say "explicitly, for the default arguments, what the input and output shapes are", I'd write what I had originally + a note that this is for the default arguments. Is that correct?

**Comment by Chillee** [comment]
I've updated the docs - I hope I interpreted your comment correctly @SsnL

**Review by soumith — APPROVED** [review]

**Comment by PistonY** [comment]
Hello, it seems that `nn.Flatten` missed in `.pyi` file.

**Comment by PistonY** [comment]
It seems that 'nn.Flatten' missed in '.pyi' file, Pycharm can't find it'

**Comment by ptrblck** [comment]
The documentation seems to be empty in `1.2.0` and `master`:
https://pytorch.org/docs/stable/nn.html#flatten

**Comment by Chillee** [comment]
@ptrblck whoops I think the `currentmodule` annotation screwed it up - I'll put up a PR fixing it.
